pytorch_lightning/utilities/auto_restart.py

```diff
@@ -0,0 +1,231 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Any, Dict, Generator, Iterator, List, Optional, Union
+
+from torch.utils.data import BatchSampler, get_worker_info, Sampler
+from torch.utils.data.dataloader import IterableDataset
+
+from pytorch_lightning.utilities.enums import AutoRestartBatchKeys
+
+
+class FastForwardSampler(Sampler):
+    """
+    This class is used to wrap a :class:`torch.utils.data.Sampler` and record the number
+    of iteration performed during an epoch.
+    On reload, if a ``state_dict`` is provided, this will be used to fast forward the wrapped samplers.
+
+    """
+
+    def __init__(self, sampler: Union[Sampler, BatchSampler, Generator]) -> None:
+        self._sampler = sampler
+        self._current_iteration = 0
+        self._dataloader_batch_size: Optional[int] = None
+        self.restarting: bool = False
+        self._cached_state_dict: Optional[Dict[str, Any]] = None
+
+    def setup(self, dataloader_batch_size: Optional[int] = None) -> None:
+        """
+        Setup the ``FastForwardSampler``.
+        This is required only when the provided dataset subclassed :class:`torch.utils.data.Dataset`.
+        """
+        self._dataloader_batch_size = dataloader_batch_size
+
+    @property
+    def worker_id(self) -> int:
+        worker_info = get_worker_info()
+        return worker_info.id if worker_info else 0
+
+    def __iter__(self) -> Iterator[List[int]]:
+        # setup local counter
+        restart_counter = 0
+
+        # iteration over wrapped sampler
+        for batch in self._sampler:
```

Comment:
```suggestion
        for batch in self.sampler:
```
Nit: Access the property directly?

Reply:
Done !